### PR TITLE
fix: prevent F.linear from saving dequantized weights in MatMul4Bit/MatMul8bitLt to save ~13GB VRAM and prevent OOM errors

### PR DESCRIPTION
**Pull request by butterwecksolutions**
## Problem                                                                                                                                                                                                                                
                                                                                                                                                                                                                                            
  During QLoRA training, `MatMul4Bit.forward` dequantizes NF4 weights                                                                                                                                                                       
  to bf16, then passes them to `F.linear()`. `F.linear` internally                                                                                                                                                                          
  creates an autograd node that saves the bf16 weight (~0.5 GB per                                                                                                                                                                          
  decoder layer for 9B models) for its backward pass. These saved                                                                                                                                                                           
  weights accumulate across all layers during forward, consuming                                                                                                                                                                            
  ~15 GB VRAM on a 32-layer model.                                                                                                                                                                                                          
                                                                                                                                                                                                                                            
  ## Root Cause                                                                                                                                                                                                                             
                                                                                                                                                                                                                                            
  `F.linear()` is a standard PyTorch op that saves its inputs for                                                                                                                                                                           
  backward via `save_for_backward`. The saved fp16 weight is redundant:
  `MatMul4Bit.backward` re-dequantizes the weight from the stored NF4                                                                                                                                                                       
  quantization state anyway (`F.dequantize_4bit(B, ctx.state)`).                                                                                                                                                                         
                                                                                                                                                                                                                                            
  ## Fix                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                            
  Wrap the `F.linear` call in `torch.no_grad()` inside both                                                                                                                                                                                 
  `MatMul4Bit.forward` and `MatMul8bitLt.forward`. This prevents the   
  intermediate autograd node from saving the dequantized weight.                                                                                                                                                                            
  Backward is unchanged — already re-dequantizes from NF4.                                                                                                                                                                                  
                                                                                                                                                                                                                                            
  **Before:**                                                                                                                                                                                                                               
                                                                                                                                                                                                                       
  output = F.linear(A, dequantize_4bit(B, state).to(A.dtype).t(), bias)                                                                                                                                                                     
                                                                                                                                                                                                                                            
  After:                                                                                                                                                                                                                                    
  with torch.no_grad():                                                                                                                                                                                                                     
      output = F.linear(A, dequantize_4bit(B, state).to(A.dtype).t(), bias)                                                                                                                                                                 
                                                                                                                                                                                                                                            
  Verification                                                                                                                                                                                                                              
                                                                                                                                                                                                                                            
  Memory snapshot analysis (torch.cuda.memory._record_memory_history)                                                                                                                                                                       
  confirmed F.linear saved fp16 weights across all decoder layers.                                                                                                                                                                          
  After the fix, only the NF4-packed weight is stored (ctx.B = uint8,                                                                                                                                                                       
  ~0.06 GB per layer).                                                                                                                                                                                                                      
                                                                                                                                                                                                                                            
                                                                                                                                                                                                                 
                                                                                                                                                                                                                                            
  | Layer | Before | After |                                                                                                                                                                                                                
  | --- | --- | --- |                                                                                                                                                                                                                       
  | Saved weight | 0.48 GB (bf16) | 0.06 GB (NF4 uint8) |                                                                                                                                                                                   
  | 32 layers total | 15.4 GB | 1.9 GB |                                                                                                                                                                                                    
                                                                                                                                                                                                                                            
  9B model training now runs on **12 GB GPUs** (was 24 GB minimum).                                                                                                                                                                         
  Zero quality impact — backward re-dequantizes to identical values.                                                                                                                                                                              

 ## Impact: Unlocked trainable Models per GPU                                                                                                                                                                                                        
                                                                                                                                                                                                                                            
  A single line change in the autograd path enables the next tier of                                                                                                                                                                        
  model sizes on every GPU class:                                      
                                                                                                                                                                                                                                            
  | GPU | VRAM | Text before | Text after | VL before | VL after |                                                                                                                                                                          
  | --- | --- | --- | --- | --- | --- |                                                                                                                                                                                                     
  | RTX 3060 | 12 GB | 4B (9B OOM) | **9B** | — | — |                                                                                                                                                                                       
  | RTX 4060 | 12 GB | 4B (9B OOM) | **9B** | — | — |                                                                                                                                                                                       
  | RTX 3080 Ti | 12 GB | 4B (9B OOM) | **9B** | — | — |                                                                                                                                                                                    
  | RTX 4070 | 12 GB | 4B (9B OOM) | **9B** | — | — |                                                                                                                                                                                       
  | RTX 5070 | 12 GB | 4B (9B OOM) | **9B** | — | — |                                                                                                                                                                                       
  | RTX 4060 Ti | 16 GB | 7B (9B OOM) | **13B** | — | **9B** |                                                                                                                                                                              
  | RTX 4080 | 16 GB | 7B (9B OOM) | **13B** | — | **9B** |                                                                                                                                                                                 
  | RTX 5080 | 16 GB | 7B (9B OOM) | **13B** | — | **9B** |                                                                                                                                                                                 
  | RTX 3090 | 24 GB | 9B (13B OOM) | **20B** | 9B (13B OOM) | **13B** |                                                                                                                                                                    
  | RTX 4090 | 24 GB | 9B (13B OOM) | **20B** | 9B (13B OOM) | **13B** |                                                                                                                                                                    
  | RTX 5090 | 32 GB | 13B (20B OOM) | **20B** | 13B (20B OOM) | **20B** |                                                                                                                                                                  
  | A10 | 24 GB | 9B (13B OOM) | **20B** | 9B (13B OOM) | **13B** |                                                                                                                                                                         
  | L40S | 48 GB | 20B (34B OOM) | **34B** | 20B (34B OOM) | **34B** |                                                                                                                                                                      
  | A100 | 80 GB | 34B (50B OOM) | **50B** | 34B (40B OOM) | **40B** |                                                                                                                                                                      
  | H100 | 80 GB | 34B (50B OOM) | **50B** | 34B (40B OOM) | **40B** |

### CODE REVIEW

**Comment by matthewdouglas** [comment]
Hi @butterwecksolutions,

Thanks for the report and the PR. I am having trouble reproducing this issue. Can you share a minimal reproducer example that shows how you're measuring memory usage, and what you're doing exactly? It may also help to know versions of PyTorch, bitsandbytes, any other relevant libraries, etc.

The way I understand it is that an `autograd.Function` runs with gradient tracking disabled already, so I'm not sure I understand how this changes anything. Perhaps you're using this with `torch.compile` and it sees the full graph with forward+backward and decides to "optimize" by reusing the dequantized weights instead of recomputing? That's the best guess I've got here.

As a side-note, I'm currently actually in the middle of doing some work around the `MatMul4Bit` autograd function, in a couple phases, and might end up turning the whole thing into a PyTorch custom op with `register_autograd` as opposed to implementing `autograd.Function`. It will be useful to have a reproducer to see if that changes anything as well.

**Comment by butterwecksolutions** [comment]
> Hi @butterwecksolutions,
> 
> Thanks for the report and the PR. I am having trouble reproducing this issue. Can you share a minimal reproducer example that shows how you're measuring memory usage, and what you're doing exactly? It may also help to know versions of PyTorch, bitsandbytes, any other relevant libraries, etc.
> 
> The way I understand it is that an `autograd.Function` runs with gradient tracking disabled already, so I'm not sure I understand how this changes anything. Perhaps you're using this with `torch.compile` and it sees the full graph with forward+backward and decides to "optimize" by reusing the dequantized weights instead of recomputing? That's the best guess I've got here.
> 
> As a side-note, I'm currently actually in the middle of doing some work around the `MatMul4Bit` autograd function, in a couple phases, and might end up turning the whole thing into a PyTorch custom op with `register_autograd` as opposed to implementing `autograd.Function`. It will be useful to have a reproducer to see if that changes anything as well.


OK, this request is a result of a debug marathon of my training pipeline. i currently build a reproducer script with all my fixes and update this within the next few days. Please wait for my results until further investigation.

**Comment by butterwecksolutions** [comment]
## Systematic Isolation Complete — Root Cause Found Elsewhere                                                                                                                                                                             
                                                                                                                                                                                                                                            
  I need to walk this PR back. After a 20+ hour training marathon with                                                                                                                                                                      
  systematic isolation across 7 patches × 2 training modes, the VRAM leak                                                                                                                                                                   
  traces to a different mechanism entirely.                                                                                                                                                                                                 
                                                                                                                                                                                                                                            
  **MatMul4Bit no_grad (#1935): 0.0 GB effect.**                                                                                                                                                                                            
                                                                                                                                                                                                                                            
  | Mode | Baseline | P1 (#1935) | Δ |                                                                                                                                                                                                      
  |------|----------|------------|---|                            
  | VL (seq_len=512) | 20.3 GB | 20.3 GB | **0.0 GB** |                                                                                                                                                                                     
  | TEXT (seq_len=4096) | 8.5 GB | 8.5 GB | **0.0 GB** |                                                                                                                                                                                    
                                                                                                                                                                                                                                            
  This PR was premature — I saw a symptom, hypothesized a cause, but hadn't                                                                                                                                                                 
  yet isolated the mechanism. The actual root cause is an async CUDA stream                                                                                                                                                                 
  leak in OffloadActivations (TRL #5700). That fix alone saves 7.0 GB —                                                                                                                                                                     
  20.3→13.3 GB in VL mode.                                                                                                                                                                                                                  
                                                                                                                                                                                                                                            
  Also tested: P2 (transformers #45769) = 0.0 GB, P5 (naive backward hook)                                                                                                                                                                  
  = OOM at 24.2 GB.                                                                                                                                                                                                                         
                                                                                                                                                                                                                                            
  **Full reproducer, HTML report, raw training logs:**                                                                                                                                                                                      
  https://github.com/butterwecksolutions/Testrepo/tree/main/vram-fix-reproducer
  https://butterwecksolutions.github.io/Testrepo/vram-fix-reproducer/vram_test_results.html                                                                                                                                                             
                                                                                                                                                                                                                                            
  Closing this PR. The real fix is in TRL #5700 — kashif has reviewed and                                                                                                                                                                   
  approved it. Thanks for your time on this.
